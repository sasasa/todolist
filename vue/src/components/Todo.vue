<template>
  <div
    class="todo"
    v-bind:class="{ 
      done: todo.finished,
      expired: isExpired(),
    }"
    v-on:click="onClick"
  >{{ todo.id }}. {{ todo.content }} 〆切({{ todo.due_on }})</div>
</template>

<script>
export default {
  name: 'Todo',
  props: {
    todo: Object
  },
  data() {
    return {
    }
  },
  computed: {
  },
  methods: {
    isExpired() {
      return Date.parse(this.todo.due_on) < Date.now()
    },
    onClick() {
      this.$emit('navigate', { todo: this.todo })
    },
  }
}
</script>

<style lang="scss" scoped>
.todo {
  display: inline-block;
  cursor: pointer;
  margin-left: 0.5em;
}
.todo:hover {
  color: #fff;
  background-color: #333;
}
.done {
  text-decoration: line-through;
}
.expired {
  color: red;
}
</style>