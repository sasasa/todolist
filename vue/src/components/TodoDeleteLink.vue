<template>
  <span
    class="cmd"
    v-show="todo.finished"
    v-on:click="onClick"
  >[X]</span>
</template>

<script>
// v-showならばテストは通るがv-ifならばテストは通らない
export default {
  name: 'TodoDeleteLink',
  props: {
    todo: Object,
  },
  data() {
    return {
    }
  },
  computed: {

  },
  methods: {
    onClick() {
      this.$emit('delete', { todo: this.todo })
    }
  }
}
</script>

<style lang="scss" scoped>
.cmd {
  cursor: pointer;
  text-decoration: underline;
  margin-left: 5px;
}
</style>