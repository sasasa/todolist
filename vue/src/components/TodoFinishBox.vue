<template>
  <input
    type="checkbox"
    v-bind:checked="todo.finished"
    v-on:click="onClick"
  />
</template>

<script>
export default {
  name: 'TodoFinishBox',
  props: {
    todo: Object
  },
  data() {
    return {
    }
  },
  computed: {

  },
  methods: {
    onClick(evt) {
      const checked = evt.target.checked;
      
      this.$emit('finish', {
        todo: {
          content: this.todo.content,
          due_on: this.todo.due_on,
          // finished: !this.todo.finished,
          finished: checked,
          id: this.todo.id,
        }
      });
    },
  },
}
</script>

<style lang="scss" scoped>

</style>