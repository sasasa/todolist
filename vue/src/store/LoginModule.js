import { buildModuleTypes } from '@/helpers/store';
// import fetch from 'cross-fetch'
import 'cross-fetch/polyfill';

const URL = process.env.VUE_APP_API_URL_BASE

const moduleName = 'LoginModule';


export const namespaced = true;

export const state = {
  errors: [
  ],
  signup_errors: {
  },
  reset_password_errors: {
  },
  signup_success: false,
  send_email_success: false,
  reset_password_success: false,
  user: {
    authentication_token: '',
    password: '',
    email: '',
    id: 0,
  },
  login: false,
};

const GETTER_TYPES = {
  USER: 'USER',
  LOGIN: 'LOGIN',
  ERRORS: 'ERRORS',
  SIGNUP_ERRORS: 'SIGNUP_ERRORS',
  SIGNUP_SUCCESS: 'SIGNUP_SUCCESS',
  SEND_EMAIL_SUCCESS: 'SEND_EMAIL_SUCCESS',
  RESET_PASSWORD_SUCCESS: 'RESET_PASSWORD_SUCCESS',
  RESET_PASSWORD_ERRORS: 'RESET_PASSWORD_ERRORS',
};
const MUTATION_TYPES = {
  USER: 'USER',
  LOGIN: 'LOGIN',
  ERRORS: 'ERRORS',
  SIGNUP_ERRORS: 'SIGNUP_ERRORS',
  SIGNUP_SUCCESS: 'SIGNUP_SUCCESS',
  SEND_EMAIL_SUCCESS: 'SEND_EMAIL_SUCCESS',
  RESET_PASSWORD_SUCCESS: 'RESET_PASSWORD_SUCCESS',
  RESET_PASSWORD_ERRORS: 'RESET_PASSWORD_ERRORS',
};
const ACTION_TYPES = {
  LOGIN_ASYNC: 'LOGIN_ASYNC',
  SIGNUP_ASYNC: 'SIGNUP_ASYNC',
  LOGOUT_ASYNC: 'LOGOUT_ASYNC',
  SEND_EMAIL_ASYNC: 'SEND_EMAIL_ASYNC',
  RESET_PASSWORD_ASYNC: 'RESET_PASSWORD_ASYNC',
};

export const LOGIN_GETTER_TYPES = buildModuleTypes({
  moduleName,
  types: GETTER_TYPES,
});
export const LOGIN_MUTATION_TYPES = buildModuleTypes({
  moduleName,
  types: MUTATION_TYPES,
});
export const LOGIN_ACTION_TYPES = buildModuleTypes({
  moduleName,
  types: ACTION_TYPES,
});

export const getters = {
  [GETTER_TYPES.USER](state) {
    return state.user
  },
  [GETTER_TYPES.LOGIN](state) {
    return state.login
  },
  [GETTER_TYPES.ERRORS](state) {
    return state.errors;
  },
  [GETTER_TYPES.SIGNUP_ERRORS](state) {
    return state.signup_errors;
  },
  [GETTER_TYPES.SIGNUP_SUCCESS](state) {
    return state.signup_success;
  },
  [GETTER_TYPES.SEND_EMAIL_SUCCESS](state) {
    return state.send_email_success
  },
  [GETTER_TYPES.RESET_PASSWORD_SUCCESS](state) {
    return state.reset_password_success
  },
  [GETTER_TYPES.RESET_PASSWORD_ERRORS](state) {
    return state.reset_password_errors
  },
};

export const mutations = {
  [MUTATION_TYPES.USER](state, { user }) {
    state.user = user
  },
  [MUTATION_TYPES.LOGIN](state, { login }) {
    state.login = login
  },
  [MUTATION_TYPES.ERRORS](state, { errors }) {
    state.errors = errors
  },
  [MUTATION_TYPES.SIGNUP_ERRORS](state, { errors }) {
    state.signup_errors = errors
  },
  [MUTATION_TYPES.SIGNUP_SUCCESS](state, { success }) {
    state.signup_success = success
  },
  [MUTATION_TYPES.SEND_EMAIL_SUCCESS](state, { success }) {
    state.send_email_success = success
  },
  [MUTATION_TYPES.RESET_PASSWORD_SUCCESS](state, { success }) {
    state.reset_password_success = success
  },
  [MUTATION_TYPES.RESET_PASSWORD_ERRORS](state, { errors }) {
    state.reset_password_errors = errors
  },
};

export const actions = {
  [ACTION_TYPES.RESET_PASSWORD_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })
    context.commit(MUTATION_TYPES.RESET_PASSWORD_SUCCESS, { success: false })

    fetch(`${URL}/users/reset_password.json`, {
      method: 'POST',
      mode: 'cors',
      credentials: 'include',
      headers: { 
        'X-CSRF-Token': localStorage.csrf,
        'Content-type': 'application/json',
      },
      body: JSON.stringify(payload),
    })
    .then(res => res.json())
    .then(data => {
      if (data.success) {
        context.commit(MUTATION_TYPES.RESET_PASSWORD_SUCCESS, { success: true })
      } else {
        context.commit(MUTATION_TYPES.RESET_PASSWORD_SUCCESS, { success: false })
        context.commit(MUTATION_TYPES.RESET_PASSWORD_ERRORS, { errors: data.errors })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error)
      alert(error)
      context.commit(MUTATION_TYPES.RESET_PASSWORD_SUCCESS, { success: false })
      context.commit('toggle_loading', payload, { root: true })
    })
  },
  [ACTION_TYPES.SEND_EMAIL_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })
    context.commit(MUTATION_TYPES.SEND_EMAIL_SUCCESS, { success: false })

    fetch(`${URL}/users/send_email.json`, {
      method: 'POST',
      mode: 'cors',
      credentials: 'include',
      headers: { 
        'X-CSRF-Token': localStorage.csrf,
        'Content-type': 'application/json',
      },
      body: JSON.stringify(payload),
    })
    .then(res => res.json())
    .then(data => {
      if (data.success) {
        context.commit(MUTATION_TYPES.SEND_EMAIL_SUCCESS, { success: true })
      } else {
        context.commit(MUTATION_TYPES.SEND_EMAIL_SUCCESS, { success: false })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error)
      alert(error)
      context.commit(MUTATION_TYPES.SEND_EMAIL_SUCCESS, { success: false })
      context.commit('toggle_loading', payload, { root: true })
    })
  },
  [ACTION_TYPES.LOGOUT_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })

    fetch(`${URL}/users/logout.json`, {
      method: 'POST',
      mode: 'cors',
      credentials: 'include',
      headers: { 
        'X-CSRF-Token': localStorage.csrf,
        'Content-type': 'application/json',
      },
      body: JSON.stringify(payload),
    })
    .then(res => res.json())
    .then(data => {
      if (data.success) {
        context.commit(MUTATION_TYPES.USER, { user: {} })
        context.commit(MUTATION_TYPES.LOGIN, { login: false })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },


  [ACTION_TYPES.LOGIN_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })

    fetch(`${URL}/users/login.json`, {
      method: 'POST',
      mode: 'cors',
      credentials: 'include',
      headers: { 'Content-type': 'application/json' },
      body: JSON.stringify(payload),
    })
    .then(res => res.json())
    .then(data => {
      if (data.errors) {
        const errors = data.errors
        context.commit(MUTATION_TYPES.ERRORS, { errors })
      } else {
        // LocalStorage に CSRF トークンを保存する
        localStorage.csrf = data.csrf;
  
        const user = data.user
        context.commit(MUTATION_TYPES.ERRORS, { errors: [] })
        context.commit(MUTATION_TYPES.USER, { user })
        context.commit(MUTATION_TYPES.LOGIN, { login: true })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },

  [ACTION_TYPES.SIGNUP_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })

    fetch(`${URL}/users.json`, {
      method: 'POST',
      mode: 'cors',
      credentials: 'include',
      headers: { 'Content-type': 'application/json' },
      body: JSON.stringify(payload),
    })
    .then(res => res.json())
    .then(data => {
      // alert(JSON.stringify(data))
      if (data.errors) {
        context.commit(MUTATION_TYPES.SIGNUP_SUCCESS, { success: false })
        context.commit(MUTATION_TYPES.SIGNUP_ERRORS, { errors: data.errors })
      } else {
        context.commit(MUTATION_TYPES.SIGNUP_SUCCESS, { success: true })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },
};