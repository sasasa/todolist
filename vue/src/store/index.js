

import Vue from 'vue';
import Vuex from 'vuex';
import createPersistedState from "vuex-persistedstate";

import * as TodoModule from './TodoModule';
import * as LoginModule from './LoginModule';

Vue.use(Vuex);

export default new Vuex.Store({
  strict: true,
  modules: {
    TodoModule,
    LoginModule,
  },
  state: {
    loading: false,
  },
  getters: {
    ['loading'](state) {
      return state.loading;
    },
  },
  mutations: {
    ['toggle_loading'](state) {
      state.loading = !state.loading 
    },
  },
  plugins: [createPersistedState()],
});

