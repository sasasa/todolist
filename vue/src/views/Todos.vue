<template>
  <div class="todos">
    <h1>Todos</h1>
    <tab-box
      v-bind:fm="fm"
      v-bind:notFinishedTodosCount="notFinishedTodos.length"
      v-bind:todosCount="todos.length"      
    />
    <transition-group tag="ul">
      <li v-for="todo of searchedTodos()" v-bind:key="todo.id">
        <todo-finish-box
          v-bind:todo="todo"
          v-on:finish="onFinish"
        />
        <todo
          v-bind:todo="todo"
          v-on:navigate="onNavigate"
        />
        <todo-delete-link
          v-bind:todo="todo"
          v-on:delete="onDelete"
        />
      </li>
    </transition-group>
  </div>
</template>


<script>
import { oneWeekBefore } from '@/helpers/date'
import Todo from '@/components/Todo'
import TodoFinishBox from '@/components/TodoFinishBox'
import TodoDeleteLink from '@/components/TodoDeleteLink'
import TabBox from '@/components/TabBox'
import { TODO_GETTER_TYPES, TODO_ACTION_TYPES } from '@/store/TodoModule';
import { LOGIN_GETTER_TYPES } from '@/store/LoginModule';
import { Titleable } from '@/mixin/titleable'

export default {
  name: 'Todos',
  mixins: [
    Titleable,
  ],
  components: {
    Todo,
    TodoFinishBox,
    TodoDeleteLink,
    TabBox,
  },
  data() {
    return {
      title: 'todolist | 一覧',

      fm: {
        expiredStr: oneWeekBefore(), // 2019-10-21
        finished: 'all',//[all, finished, not_finished]
        activeTab: 'A',//[A, B]
      },
    }
  },
  mounted() {
    return this.$store.dispatch(TODO_ACTION_TYPES.GET_TODOS_ASYNC, { user: this.user })
  },
  computed: {
    user() {
      return this.$store.getters[LOGIN_GETTER_TYPES.USER]
    },
    todos() {
      return this.$store.getters[TODO_GETTER_TYPES.TODOS]
    },
    sortedTodos() {
      return this.todos.slice().sort((a, b) => { 
        return a.id < b.id ? 1 : -1
      })
    },
    notFinishedTodos() {
      return this.sortedTodos.filter((todo) => {
        return !todo.finished
      })
    },
    finishedTodos() {
      return this.sortedTodos.filter((todo) => {
        return todo.finished
      })
    },
    timeSortedTodos() {
      return this.todos.slice().sort((a, b) => { 
        return Date.parse(a.due_on) > Date.parse(b.due_on) ? 1 : -1
      })
    },
  },
  methods: {
    searchedTodos() {
      if (this.fm.activeTab === 'A') {
        switch (this.fm.finished) {
          case 'all':
            return this.sortedTodos
          case 'not_finished':
            return this.notFinishedTodos
          case 'finished':
            return this.finishedTodos
        }
      } else {
        return this.getTodosByTime(this.fm.expiredStr)
      }
    },
    getTodosByTime(timeStr) {
      let compared = Date.parse(timeStr)
      if (compared) {
        return this.timeSortedTodos.filter((todo) => {
          return Date.parse(todo.due_on) >= compared
        })
      } else {
        return this.timeSortedTodos
      }
    },
    onFinish({ todo }) {
      return this.$store.dispatch(TODO_ACTION_TYPES.UPDATE_TODO_ASYNC, { todo, user: this.user })
    },
    onNavigate({ todo }) {
      return this.$router.push(`/edit/${todo.id}`)
    },
    onDelete({ todo }) {
      if (confirm("削除しますか？")) {
        return this.$store.dispatch(TODO_ACTION_TYPES.REMOVE_TODO_ASYNC, { todo, user: this.user })
      }
    },
  },
}
</script>


<style lang="scss" scoped>
.v-enter-active,
.v-leave-active,
.v-move {
  transition: all .5s;
}
.v-leave-active {
  position: absolute;
}
.v-enter,
.v-leave-to {
  transform: translateX(40%) rotate(-6deg);
  opacity: 0;
}
ul {
  padding: 0;
  margin: 0;
  list-style: none;
}

</style>