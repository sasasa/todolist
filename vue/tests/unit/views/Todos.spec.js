import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import Vuex from 'vuex'
import Todos from '@/views/Todos'
import * as TodoModule from '../store/TodoModuleMock'
import Todo from '@/components/Todo'
import TodoFinishBox from '@/components/TodoFinishBox'
import TodoDeleteLink from '@/components/TodoDeleteLink'

const localVue = createLocalVue()
// localVue.use(VueRouter)
localVue.use(Vuex)

const router = new VueRouter()

describe('Todos view', () => {
  let store
  let $router
  let $store
  let todo

  beforeEach(() => {
    store = new Vuex.Store({
      modules: {
        TodoModule,
      }
    })
    $router = {
      push: jest.fn()
    }
    todo = {
      id: 1,
      content: 'test content',
    }
  })

  it('mount', () => {
    const wrapper = shallowMount(Todos, {
      store,
      router,
      localVue,
    })
    expect.assertions(2)
    expect(wrapper.isVueInstance()).toBe(true)
    // このテストは一番最初の it でないと通らないのでテストの順番を変えてはいけない
    expect(TodoModule.actions[TodoModule.ACTION_TYPES.GET_TODOS_ASYNC].mock.calls).toHaveLength(1)
  })
  describe('computed', () => {
    it('notFinishedTodos', () => {
      const wrapper = shallowMount(Todos, {
        store,
        router,
        localVue,
      })
      expect(wrapper.vm.notFinishedTodos).toEqual([
        {
          "id": 4,
          "content": "Vue.jsの勉強をする!!",
          "finished": false,
          "due_on": "2019-10-03"
        },
        {
          "id": 2,
          "content": "3キロ走る",
          "finished": false,
          "due_on": "2019-10-04"
        },
        {
          "id": 1,
          "content": "ちょっと筋トレする",
          "finished": false,
          "due_on": "2019-10-06"
        },
      ])
    })
    it('finishedTodos', () => {
      const wrapper = shallowMount(Todos, {
        store,
        router,
        localVue,
      })
      expect(wrapper.vm.finishedTodos).toEqual([
        {
          "content": "運動不足を解消する",
          "due_on": "2019-10-14",
          "finished": true,
          "id": 5
        },
        {
          "id": 3,
          "content": "牛乳を買う",
          "finished": true,
          "due_on": "2019-10-08"
        },
      ])
    })
    it('timeSortedTodos', () => {
      const wrapper = shallowMount(Todos, {
        store,
        router,
        localVue,
      })
      expect(wrapper.vm.timeSortedTodos).toEqual([
        {
          "id": 4,
          "content": "Vue.jsの勉強をする!!",
          "finished": false,
          "due_on": "2019-10-03"
        },
        {
          "id": 2,
          "content": "3キロ走る",
          "finished": false,
          "due_on": "2019-10-04"
        },
        {
          "id": 1,
          "content": "ちょっと筋トレする",
          "finished": false,
          "due_on": "2019-10-06"
        },
        {
          "id": 3,
          "content": "牛乳を買う",
          "finished": true,
          "due_on": "2019-10-08"
        },
        {
          "content": "運動不足を解消する",
          "due_on": "2019-10-14",
          "finished": true,
          "id": 5
        },
      ])
    })

    it('sortedTodos', () => {
      const wrapper = shallowMount(Todos, {
        store,
        router,
        localVue,
      })
      expect(wrapper.vm.sortedTodos).toEqual([
        {
          "content": "運動不足を解消する",
          "due_on": "2019-10-14",
          "finished": true,
          "id": 5
        },
        {
          "id": 4,
          "content": "Vue.jsの勉強をする!!",
          "finished": false,
          "due_on": "2019-10-03"
        },
        {
          "id": 3,
          "content": "牛乳を買う",
          "finished": true,
          "due_on": "2019-10-08"
        },
        {
          "id": 2,
          "content": "3キロ走る",
          "finished": false,
          "due_on": "2019-10-04"
        },
        {
          "id": 1,
          "content": "ちょっと筋トレする",
          "finished": false,
          "due_on": "2019-10-06"
        },
      ])
    })

  })
  describe('methods', () => {
    describe('searchedTodos', () => {
      describe('fm.activeTab === "A" のとき', () => {
        it('fm.finished === "all" のとき', () => {
          const wrapper = shallowMount(Todos, {
            store,
            router,
            localVue,
          })
          wrapper.vm.fm.activeTab = 'A'
          wrapper.vm.fm.finished = 'all'
          expect(wrapper.vm.searchedTodos()).toEqual(wrapper.vm.sortedTodos)
        })
        it('fm.finished === "not_finished" のとき', () => {
          const wrapper = shallowMount(Todos, {
            store,
            router,
            localVue,
          })
          wrapper.vm.fm.activeTab = 'A'
          wrapper.vm.fm.finished = 'not_finished'
          expect(wrapper.vm.searchedTodos()).toEqual(wrapper.vm.notFinishedTodos)
        })
        it('fm.finished === "finished" のとき', () => {
          const wrapper = shallowMount(Todos, {
            store,
            router,
            localVue,
          })
          wrapper.vm.fm.activeTab = 'A'
          wrapper.vm.fm.finished = 'finished'
          expect(wrapper.vm.searchedTodos()).toEqual(wrapper.vm.finishedTodos)
        })
      })
      it('fm.activeTab === "B" のとき', () => {
        const mock = jest.fn()
        const wrapper = shallowMount(Todos, {
          store,
          router,
          localVue,
        })
        wrapper.setMethods({
          getTodosByTime: mock
        })
        wrapper.vm.fm.activeTab = 'B'
        wrapper.vm.searchedTodos()
        expect(mock).toHaveBeenCalled()
      })
    })

    describe('getTodosByTime(timeStr)', () => {
      it('timeStr="2019-10-07" のとき', () => {
        const wrapper = shallowMount(Todos, {
          store,
          router,
          localVue,
        })
        expect(wrapper.vm.getTodosByTime('2019-10-07')).toEqual([
          {
            "id": 3,
            "content": "牛乳を買う",
            "finished": true,
            "due_on": "2019-10-08"
          },
          {
            "content": "運動不足を解消する",
            "due_on": "2019-10-14",
            "finished": true,
            "id": 5
          },
        ])
      })
      it('timeStr="" のとき', () => {
        const wrapper = shallowMount(Todos, {
          store,
          router,
          localVue,
        })
        expect(wrapper.vm.getTodosByTime('')).toEqual([
          {
            "id": 4,
            "content": "Vue.jsの勉強をする!!",
            "finished": false,
            "due_on": "2019-10-03"
          },
          {
            "id": 2,
            "content": "3キロ走る",
            "finished": false,
            "due_on": "2019-10-04"
          },
          {
            "id": 1,
            "content": "ちょっと筋トレする",
            "finished": false,
            "due_on": "2019-10-06"
          },
          {
            "id": 3,
            "content": "牛乳を買う",
            "finished": true,
            "due_on": "2019-10-08"
          },
          {
            "content": "運動不足を解消する",
            "due_on": "2019-10-14",
            "finished": true,
            "id": 5
          },
        ])
      })
    })

    it('onNavigate', () => {
      const wrapper = shallowMount(Todos, {
        store,
        router,
        localVue,
        mocks: {
          $router,
        }
      })

      wrapper.vm.onNavigate({ todo })
      expect.assertions(2)
      expect($router.push.mock.calls).toHaveLength(1)
      expect($router.push).toHaveBeenCalledWith(`/edit/${todo.id}`)
    })

    it('onFinish', () => {
      const wrapper = shallowMount(Todos, {
        store,
        router,
        localVue,
        mocks: {
          $router,
        }
      })

      wrapper.vm.onFinish({ todo })
      expect(TodoModule.actions[TodoModule.ACTION_TYPES.UPDATE_TODO_ASYNC].mock.calls).toHaveLength(1)
      expect(TodoModule.actions[TodoModule.ACTION_TYPES.UPDATE_TODO_ASYNC].mock.calls[0][1]).toEqual({ todo })
    })

    describe('onDelete', () => {
      describe('confirm false の時', () => {
        beforeEach(() => {
          window.confirm = jest.fn(() => false)
        })
        it('メソッドが呼び出されない', () => {
          const wrapper = shallowMount(Todos, {
            store,
            router,
            localVue,
            mocks: {
              $router,
            }
          })
          wrapper.vm.onDelete({ todo })
          expect(window.confirm).toBeCalled()
          expect(TodoModule.actions[TodoModule.ACTION_TYPES.REMOVE_TODO_ASYNC]).not.toBeCalled()
        })
      })
      describe('confirm true の時', () => {
        beforeEach(() => {
          window.confirm = jest.fn(() => true)
        })
        it('メソッドが呼び出される', () => {
          const wrapper = shallowMount(Todos, {
            store,
            router,
            localVue,
            mocks: {
              $router,
            }
          })
          wrapper.vm.onDelete({ todo })
          expect(window.confirm).toBeCalled()
          expect(TodoModule.actions[TodoModule.ACTION_TYPES.REMOVE_TODO_ASYNC].mock.calls).toHaveLength(1)
          expect(TodoModule.actions[TodoModule.ACTION_TYPES.REMOVE_TODO_ASYNC].mock.calls[0][1]).toEqual({ todo })
        })
      })
    })

  })

  describe('template', () => {
    it('@finish=onFinish', () => {
      const mock = jest.fn()
      const wrapper = shallowMount(Todos, {
        store,
        router,
        localVue,
        mocks: {
          $router,
        }
      })
      wrapper.setMethods({
        onFinish: mock
      })
      const todoFinishBox = wrapper.find(TodoFinishBox)
      todoFinishBox.vm.$emit('finish')
      expect(mock).toHaveBeenCalled()
    })

    it('@navigate=onNavigate', () => {
      const mock = jest.fn()
      const wrapper = shallowMount(Todos, {
        store,
        router,
        localVue,
        mocks: {
          $router,
        }
      })
      wrapper.setMethods({
        onNavigate: mock
      })
      const t = wrapper.find(Todo)
      t.vm.$emit('navigate')
      expect(mock).toHaveBeenCalled()
    })

    it('@delete=onDelete', () => {
      const mock = jest.fn()
      const wrapper = shallowMount(Todos, {
        store,
        router,
        localVue,
        mocks: {
          $router,
        }
      })
      wrapper.setMethods({
        onDelete: mock
      })
      const todoDeleteLink = wrapper.find(TodoDeleteLink)
      todoDeleteLink.vm.$emit('delete')
      expect(mock).toHaveBeenCalled()
    })


    it('snapshot', () => {
      const wrapper = shallowMount(Todos, {
        store,
        router,
        localVue,
        mocks: {
          $router,
          $store,
        }
      })
      expect(wrapper.vm.$el).toMatchSnapshot()
    })
  })
})