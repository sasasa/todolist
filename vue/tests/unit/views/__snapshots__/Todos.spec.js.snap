// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos view template snapshot 1`] = `
<div
  class="todos"
>
  <h1>
    Todos
  </h1>
   
  <tab-box-stub
    fm="[object Object]"
    notfinishedtodoscount="3"
    todoscount="5"
  />
   
  <span
    tag="ul"
  >
    <li>
      <todo-finish-box-stub
        todo="[object Object]"
      />
       
      <todo-stub
        todo="[object Object]"
      />
       
      <todo-delete-link-stub
        todo="[object Object]"
      />
    </li>
    <li>
      <todo-finish-box-stub
        todo="[object Object]"
      />
       
      <todo-stub
        todo="[object Object]"
      />
       
      <todo-delete-link-stub
        todo="[object Object]"
      />
    </li>
    <li>
      <todo-finish-box-stub
        todo="[object Object]"
      />
       
      <todo-stub
        todo="[object Object]"
      />
       
      <todo-delete-link-stub
        todo="[object Object]"
      />
    </li>
    <li>
      <todo-finish-box-stub
        todo="[object Object]"
      />
       
      <todo-stub
        todo="[object Object]"
      />
       
      <todo-delete-link-stub
        todo="[object Object]"
      />
    </li>
    <li>
      <todo-finish-box-stub
        todo="[object Object]"
      />
       
      <todo-stub
        todo="[object Object]"
      />
       
      <todo-delete-link-stub
        todo="[object Object]"
      />
    </li>
  </span>
</div>
`;
